Pipeline with single-end reads sample as an input

I'm trying to get a pipeline working with single-end reads sample as an input. The pipeline has been defined but when a single-end reads sample is added to the cart and the pipeline is selected, the pipelines page doesn't load correctly and the catalina.out file logs the following error:

ERROR org.thymeleaf.TemplateEngine:1136 - [THYMELEAF][http-bio-8080-exec-4] Exception processing template "pipelines/types/generic_pipeline": Exception evaluating SpringEL expression: "join.getObject()" (template: "pipelines/types/generic_pipeline" - line 265, col 53)
org.thymeleaf.exceptions.TemplateProcessingException: Exception evaluating SpringEL expression: "join.getObject()" (template: "pipelines/types/generic_pipeline" - line 265, col 53)
....
Caused by: org.springframework.expression.spel.SpelEvaluationException: EL1004E:(pos 5): Method call: Method getObject() cannot be found on ca.corefacility.bioinformatics.irida.model.sequenceFile.SingleEndSequenceFile type

When a paired-end reads sample is added to the cart and the same pipeline selected, the pipeline page loads fine but obviously the message "This sample does not contain any files that can be run on this pipeline." is shown.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,7 @@ Changes
 20.01 to 20.05
 --------------
 * [UI]: Fixed bug that prevented server side filtering for analysis name on analyses table. (20.01.01)
+* [UI]: Fixed UI bugs preventing users from selecting single-end input files for analysis pipelines. (20.01.2)
 
 19.09 to 20.01
 --------------

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
 	<groupId>ca.corefacility.bioinformatics</groupId>
 	<artifactId>irida</artifactId>
 	<packaging>war</packaging>
-	<version>20.01.01</version>
+	<version>20.01.2</version>
 	<name>irida</name>
 	<url>http://www.irida.ca</url>
 

--- a/src/main/java/ca/corefacility/bioinformatics/irida/ria/web/analysis/AnalysisController.java
+++ b/src/main/java/ca/corefacility/bioinformatics/irida/ria/web/analysis/AnalysisController.java
@@ -1,27 +1,5 @@
 package ca.corefacility.bioinformatics.irida.ria.web.analysis;
 
-import java.io.*;
-import java.nio.file.Files;
-import java.nio.file.Path;
-import java.security.Principal;
-import java.util.*;
-import java.util.stream.Collectors;
-
-import javax.servlet.http.HttpServletResponse;
-
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.context.MessageSource;
-import org.springframework.context.annotation.Scope;
-import org.springframework.http.MediaType;
-import org.springframework.security.access.prepost.PreAuthorize;
-import org.springframework.security.core.Authentication;
-import org.springframework.security.core.context.SecurityContextHolder;
-import org.springframework.stereotype.Controller;
-import org.springframework.ui.Model;
-import org.springframework.web.bind.annotation.*;
-
 import ca.corefacility.bioinformatics.irida.exceptions.EntityNotFoundException;
 import ca.corefacility.bioinformatics.irida.exceptions.ExecutionManagerException;
 import ca.corefacility.bioinformatics.irida.exceptions.IridaWorkflowNotFoundException;
@@ -35,6 +13,7 @@ import ca.corefacility.bioinformatics.irida.model.sample.Sample;
 import ca.corefacility.bioinformatics.irida.model.sample.SampleSequencingObjectJoin;
 import ca.corefacility.bioinformatics.irida.model.sample.metadata.MetadataEntry;
 import ca.corefacility.bioinformatics.irida.model.sequenceFile.SequenceFilePair;
+import ca.corefacility.bioinformatics.irida.model.sequenceFile.SingleEndSequenceFile;
 import ca.corefacility.bioinformatics.irida.model.user.Role;
 import ca.corefacility.bioinformatics.irida.model.user.User;
 import ca.corefacility.bioinformatics.irida.model.workflow.IridaWorkflow;
@@ -57,7 +36,6 @@ import ca.corefacility.bioinformatics.irida.service.sample.MetadataTemplateServi
 import ca.corefacility.bioinformatics.irida.service.sample.SampleService;
 import ca.corefacility.bioinformatics.irida.service.user.UserService;
 import ca.corefacility.bioinformatics.irida.service.workflow.IridaWorkflowsService;
-
 import com.fasterxml.jackson.core.JsonParseException;
 import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.JsonMappingException;
@@ -66,6 +44,26 @@ import com.google.common.base.Strings;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.MessageSource;
+import org.springframework.context.annotation.Scope;
+import org.springframework.http.MediaType;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.*;
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.Principal;
+import java.util.*;
+import java.util.stream.Collectors;
 
 /**
  * Controller for Analysis.
@@ -276,6 +274,22 @@ public class AnalysisController {
 					.compareTo(b.sample.getLabel());
 		}).collect(Collectors.toList());
 		model.addAttribute("paired_end", sampleFiles);
+
+		// - Single
+		Set<SingleEndSequenceFile> inputFilesSingle = sequencingObjectService.getSequencingObjectsOfTypeForAnalysisSubmission(
+				submission, SingleEndSequenceFile.class);
+		List<SampleSingleFiles> singleFiles = inputFilesSingle.stream().map(SampleSingleFiles::new).sorted((a, b) -> {
+			if (a.sample == null && b.sample == null) {
+				return 0;
+			} else if (a.sample == null) {
+				return -1;
+			} else if (b.sample == null) {
+				return 1;
+			}
+			return a.sample.getLabel()
+					.compareTo(b.sample.getLabel());
+		}).collect(Collectors.toList());
+		model.addAttribute("single_end", singleFiles);
 
 		// Check if user can update analysis
 		Authentication authentication = SecurityContextHolder.getContext()
@@ -1112,6 +1126,39 @@ public class AnalysisController {
 
 		public SequenceFilePair getSequenceFilePair() {
 			return sequenceFilePair;
+		}
+	}
+
+	/**
+	 * UI Model to return a single end sequence file with its sample
+	 */
+	private class SampleSingleFiles {
+		private Sample sample;
+		private SingleEndSequenceFile singleEndFile;
+
+		SampleSingleFiles(SingleEndSequenceFile singleEndFile) {
+			this.singleEndFile = singleEndFile;
+			try {
+				SampleSequencingObjectJoin sampleSequencingObjectJoin = sampleService.getSampleForSequencingObject(
+						singleEndFile);
+				this.sample = sampleSequencingObjectJoin.getSubject();
+			} catch (Exception e) {
+				logger.debug(
+						"Sequence file [" + singleEndFile.getIdentifier() + "] does not have a parent sample", e);
+				sample = null;
+			}
+		}
+
+		public Long getId() {
+			return singleEndFile.getId();
+		}
+
+		public Sample getSample() {
+			return sample;
+		}
+
+		public SingleEndSequenceFile getSequenceFile() {
+			return singleEndFile;
 		}
 	}
 }

--- a/src/main/java/ca/corefacility/bioinformatics/irida/ria/web/analysis/AnalysisController.java
+++ b/src/main/java/ca/corefacility/bioinformatics/irida/ria/web/analysis/AnalysisController.java
@@ -13,6 +13,7 @@ import ca.corefacility.bioinformatics.irida.model.sample.Sample;
 import ca.corefacility.bioinformatics.irida.model.sample.SampleSequencingObjectJoin;
 import ca.corefacility.bioinformatics.irida.model.sample.metadata.MetadataEntry;
 import ca.corefacility.bioinformatics.irida.model.sequenceFile.SequenceFilePair;
+import ca.corefacility.bioinformatics.irida.model.sequenceFile.SequencingObject;
 import ca.corefacility.bioinformatics.irida.model.sequenceFile.SingleEndSequenceFile;
 import ca.corefacility.bioinformatics.irida.model.user.Role;
 import ca.corefacility.bioinformatics.irida.model.user.User;
@@ -262,33 +263,19 @@ public class AnalysisController {
 		// - Paired
 		Set<SequenceFilePair> inputFilePairs = sequencingObjectService.getSequencingObjectsOfTypeForAnalysisSubmission(
 				submission, SequenceFilePair.class);
-		List<SampleFiles> sampleFiles = inputFilePairs.stream().map(SampleFiles::new).sorted((a, b) -> {
-			if (a.sample == null && b.sample == null) {
-				return 0;
-			} else if (a.sample == null) {
-				return -1;
-			} else if (b.sample == null) {
-				return 1;
-			}
-			return a.sample.getLabel()
-					.compareTo(b.sample.getLabel());
-		}).collect(Collectors.toList());
+		List<SampleSequencingObject> sampleFiles = inputFilePairs.stream()
+				.map(SampleSequencingObject::new)
+				.sorted()
+				.collect(Collectors.toList());
 		model.addAttribute("paired_end", sampleFiles);
 
 		// - Single
 		Set<SingleEndSequenceFile> inputFilesSingle = sequencingObjectService.getSequencingObjectsOfTypeForAnalysisSubmission(
 				submission, SingleEndSequenceFile.class);
-		List<SampleSingleFiles> singleFiles = inputFilesSingle.stream().map(SampleSingleFiles::new).sorted((a, b) -> {
-			if (a.sample == null && b.sample == null) {
-				return 0;
-			} else if (a.sample == null) {
-				return -1;
-			} else if (b.sample == null) {
-				return 1;
-			}
-			return a.sample.getLabel()
-					.compareTo(b.sample.getLabel());
-		}).collect(Collectors.toList());
+		List<SampleSequencingObject> singleFiles = inputFilesSingle.stream()
+				.map(SampleSequencingObject::new)
+				.sorted()
+				.collect(Collectors.toList());
 		model.addAttribute("single_end", singleFiles);
 
 		// Check if user can update analysis
@@ -1097,68 +1084,48 @@ public class AnalysisController {
 	}
 
 	/**
-	 * UI Model to return a pair aof Sequence files with its accompanying sample.
+	 * UI Model to return Sequence files with its accompanying sample.
 	 */
-	private class SampleFiles {
+	private class SampleSequencingObject implements Comparable<SampleSequencingObject>{
 		private Sample sample;
-		private SequenceFilePair sequenceFilePair;
+		private SequencingObject sequencingObject;
 
-		SampleFiles(SequenceFilePair sequenceFilePair) {
-			this.sequenceFilePair = sequenceFilePair;
+		SampleSequencingObject(SequencingObject sequencingObject) {
+			this.sequencingObject = sequencingObject;
 			try {
 				SampleSequencingObjectJoin sampleSequencingObjectJoin = sampleService.getSampleForSequencingObject(
-						sequenceFilePair);
+						sequencingObject);
 				this.sample = sampleSequencingObjectJoin.getSubject();
 			} catch (Exception e) {
-				logger.debug(
-						"Sequence file pair [" + sequenceFilePair.getIdentifier() + "] does not have a parent sample", e);
+				logger.debug("Sequence file [" + sequencingObject.getIdentifier() + "] does not have a parent sample",
+						e);
 				sample = null;
 			}
 		}
 
 		public Long getId() {
-			return sequenceFilePair.getId();
+			return sequencingObject.getId();
 		}
 
 		public Sample getSample() {
 			return sample;
 		}
 
-		public SequenceFilePair getSequenceFilePair() {
-			return sequenceFilePair;
+		public SequencingObject getSequencingObject() {
+			return sequencingObject;
 		}
-	}
 
-	/**
-	 * UI Model to return a single end sequence file with its sample
-	 */
-	private class SampleSingleFiles {
-		private Sample sample;
-		private SingleEndSequenceFile singleEndFile;
-
-		SampleSingleFiles(SingleEndSequenceFile singleEndFile) {
-			this.singleEndFile = singleEndFile;
-			try {
-				SampleSequencingObjectJoin sampleSequencingObjectJoin = sampleService.getSampleForSequencingObject(
-						singleEndFile);
-				this.sample = sampleSequencingObjectJoin.getSubject();
-			} catch (Exception e) {
-				logger.debug(
-						"Sequence file [" + singleEndFile.getIdentifier() + "] does not have a parent sample", e);
-				sample = null;
+		@Override
+		public int compareTo(SampleSequencingObject b) {
+			if (this.sample == null && b.sample == null) {
+				return 0;
+			} else if (this.sample == null) {
+				return -1;
+			} else if (b.sample == null) {
+				return 1;
 			}
-		}
-
-		public Long getId() {
-			return singleEndFile.getId();
-		}
-
-		public Sample getSample() {
-			return sample;
-		}
-
-		public SingleEndSequenceFile getSequenceFile() {
-			return singleEndFile;
+			return this.sample.getLabel()
+					.compareTo(b.sample.getLabel());
 		}
 	}
 }

--- a/src/main/webapp/pages/analysis/details/details.html
+++ b/src/main/webapp/pages/analysis/details/details.html
@@ -361,7 +361,7 @@
 							  <span th:text="#{analysis.details.input.no-sample}">_NO_ONE_HOME_</span>
 						  </h4>
               <hr style="margin: 10px 0">
-              <div th:with="forward=${pair.sequenceFilePair.getForwardSequenceFile()}" class="sequence-file">
+              <div th:with="forward=${pair.getSequencingObject().getForwardSequenceFile()}" class="sequence-file">
                 <i class="fas fa-arrow-right sequence-file__icon"></i> 
                 <span class="sequence-file__name">
                   <a th:href="@{/sequenceFiles/{id}/file/{fileId}/summary(id=${pair.id}, fileId=${forward.id})}"
@@ -369,7 +369,7 @@
                 </span>
                 <span th:text="${forward.getFileSize()}"></span>
               </div>
-              <div th:with="reverse=${pair.sequenceFilePair.getReverseSequenceFile()}" class="sequence-file">
+              <div th:with="reverse=${pair.getSequencingObject().getReverseSequenceFile()}" class="sequence-file">
                 <i class="fas fa-arrow-left sequence-file__icon"></i>
                 <span class="sequence-file__name">
                   <a th:href="@{/sequenceFiles/{id}/file/{fileId}/summary(id=${pair.id}, fileId=${reverse.id})}"
@@ -393,7 +393,7 @@
 							  <span th:text="#{analysis.details.input.no-sample}">_NO_ONE_HOME_</span>
 						  </h4>
               <hr style="margin: 10px 0">
-              <div th:with="forward=${single.getSequenceFile().getSequenceFile()}" class="sequence-file">
+              <div th:with="forward=${single.getSequencingObject().getSequenceFile()}" class="sequence-file">
                 <i class="fas fa-minus sequence-file__icon"></i> 
                 <span class="sequence-file__name">
                   <a th:href="@{/sequenceFiles/{id}/file/{fileId}/summary(id=${single.id}, fileId=${forward.id})}"

--- a/src/main/webapp/pages/analysis/details/details.html
+++ b/src/main/webapp/pages/analysis/details/details.html
@@ -334,7 +334,7 @@
           <script
             type="text/ng-template"
             id="inputs.html"
-            th:with="fileCount = ${#lists.size(paired_end)}"
+            th:with="pairedCount = ${#lists.size(paired_end)}, singleCount = ${#lists.size(single_end)}"
           >
 			  <th:block th:if="${referenceFile}">
 				  <h2 th:text="#{analysis.details.reference}"></h2>
@@ -347,7 +347,7 @@
 			  </th:block>
 
 			  <h2 th:text="#{analysis.details.reads}"></h2>
-			  <div th:if="${fileCount > 0}">
+			  <div th:if="${pairedCount > 0}">
 				  <ul class="list-group">
 					  <li class="list-group-item t-paired-end" th:each="pair : ${paired_end}">
 						  <h4 class="list-group-item-heading" th:if="${pair.sample}">
@@ -360,26 +360,54 @@
 							  <i class="far fa-times-circle"></i>
 							  <span th:text="#{analysis.details.input.no-sample}">_NO_ONE_HOME_</span>
 						  </h4>
-            <hr style="margin: 10px 0">
-                     			<div th:with="forward=${pair.sequenceFilePair.getForwardSequenceFile()}" class="sequence-file">
-                     				<i class="fas fa-arrow-right sequence-file__icon"></i> <span class="sequence-file__name"><a
-                 th:href="@{/sequenceFiles/{id}/file/{fileId}/summary(id=${pair.id}, fileId=${forward.id})}"
-                 th:text="${forward.label}"></a></span>
-                     				<span th:text="${forward.getFileSize()}"></span>
-                     			</div>
-                     			<div th:with="reverse=${pair.sequenceFilePair.getReverseSequenceFile()}" class="sequence-file">
-                     				<i class="fas fa-arrow-left sequence-file__icon"></i> <span class="sequence-file__name"><a
-                 th:href="@{/sequenceFiles/{id}/file/{fileId}/summary(id=${pair.id}, fileId=${reverse.id})}"
-                 th:text="${reverse.label}"></a></span> <span th:text="${reverse.getFileSize()}"></span>
-                     			</div>
-                     		</li>
-                     	</ul>
-                     </div>
-                     <div class="alert alert-info" th:unless="${fileCount > 0}"
-                     	 th:text="#{analysis.input-files.no-paired-end}">
-                     	__NO_PAIRED_END_DATA_
-                     </div>
-          </script>
+              <hr style="margin: 10px 0">
+              <div th:with="forward=${pair.sequenceFilePair.getForwardSequenceFile()}" class="sequence-file">
+                <i class="fas fa-arrow-right sequence-file__icon"></i> 
+                <span class="sequence-file__name">
+                  <a th:href="@{/sequenceFiles/{id}/file/{fileId}/summary(id=${pair.id}, fileId=${forward.id})}"
+                    th:text="${forward.label}"></a>
+                </span>
+                <span th:text="${forward.getFileSize()}"></span>
+              </div>
+              <div th:with="reverse=${pair.sequenceFilePair.getReverseSequenceFile()}" class="sequence-file">
+                <i class="fas fa-arrow-left sequence-file__icon"></i>
+                <span class="sequence-file__name">
+                  <a th:href="@{/sequenceFiles/{id}/file/{fileId}/summary(id=${pair.id}, fileId=${reverse.id})}"
+                    th:text="${reverse.label}"></a>
+                </span> <span th:text="${reverse.getFileSize()}"></span>
+              </div>
+            </li>
+          </ul>
+        </div>
+        <div th:if="${singleCount > 0}">
+          <ul class="list-group">
+					  <li class="list-group-item t-single-end" th:each="single : ${single_end}">
+						  <h4 class="list-group-item-heading" th:if="${single.sample}">
+							  <a target="_blank" th:href="@{/samples/{id}/details(id=${single.sample.id})}">
+								  <i class="fas fa-flask"></i>
+								  <span th:text="${single.sample.sampleName}"></span>
+							  </a>
+						  </h4>
+						  <h4 th:unless="${single.sample}">
+							  <i class="far fa-times-circle"></i>
+							  <span th:text="#{analysis.details.input.no-sample}">_NO_ONE_HOME_</span>
+						  </h4>
+              <hr style="margin: 10px 0">
+              <div th:with="forward=${single.getSequenceFile().getSequenceFile()}" class="sequence-file">
+                <i class="fas fa-minus sequence-file__icon"></i> 
+                <span class="sequence-file__name">
+                  <a th:href="@{/sequenceFiles/{id}/file/{fileId}/summary(id=${single.id}, fileId=${forward.id})}"
+                    th:text="${forward.label}"></a>
+                </span>
+                <span th:text="${forward.getFileSize()}"></span>
+              </div>
+            </li>
+          </ul>
+        </div>
+        <div class="alert alert-info" th:unless="${pairedCount > 0 OR singleCount > 0}"
+          th:text="#{analysis.input-files.no-paired-end}">__NO_PAIRED_END_DATA_
+        </div>
+      </script>
 
           <script type="text/ng-template" id="provenance.html">
             <div th:if="${isCompleted}" th:with="files=${analysis.getAnalysisOutputFiles()}">

--- a/src/main/webapp/pages/pipelines/types/generic_pipeline.html
+++ b/src/main/webapp/pages/pipelines/types/generic_pipeline.html
@@ -278,16 +278,15 @@
                                             </th:block>
 
                                             <!--/* (15-03-26 ) SINGLE END DATA */-->
-                                            <th:block th:if="${hasSingles}" th:with="files=${sample.files.single_end}">
-                                                <tr class="single_end" th:each="join : ${files}"
-                                                    th:with="file=${join.getObject()}">
+                                            <th:block th:if="${hasSingles}" th:with="singles=${sample.files.single_end}">
+                                                <tr class="single_end" th:each="file : ${singles}">
                                                     <td>
                                                         <!--/* (15-03-26 ) Checked only if there are no pairs and it is the first one in the file set */-->
                                                         <input type="radio" data-type="single_end"
                                                                th:value="${file.getId()}" th:name="${sample.id}"
                                                                th:checked="${!hasPaired and fileStat.first}"/>
                                                     </td>
-                                                    <td th:replace="templates/_sequenceFile :: sequenceFileTableLayout (object=${pair},file=${file},icon='single',relative='false')"></td>
+                                                    <td th:replace="templates/_sequenceFile :: sequenceFileTableLayout (object=${file},file=${file.getSequenceFile()},icon='single',relative='false')"></td>
                                                 </tr>
                                             </th:block>
                                         </table>
@@ -343,7 +342,7 @@
                             th:href="@{/analysis}"
                             th:text="#{workflow.submitted-view-btn}">Let's see how this pipeline is doing.
                             </a>
-                            <br/></br/>
+                            <br/><br/>
                             <button id="btn-clear-pipeline" class="btn btn-warning btn-affixed"
                                     ng-click="pipeline.clearAndRedirect()"
                                     th:text="#{workflow.submitted-clear-btn}">Clear the cart and start over!

